### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by bkirwi** [mslinn/decline-test#1]
https://github.com/mslinn/decline-test/blob/3dce3b03e60972d36830865653f334b014f6d46d/src/main/scala/Cad.scala#L155

Options and flags are also not optional unless you provide an alternative or default value, a la: http://localhost:4000/decline/usage.html#default-values. IMO using positional arguments for mandatory things is more idiomatic, though.

https://github.com/mslinn/decline-test/blob/3dce3b03e60972d36830865653f334b014f6d46d/src/main/scala/Cad.scala#L200

It sounds like you want usage that looks like `cad --debug subcommand [...]`... having options both at the top level and in the subcommands. The usual way to get this is to define your subcommands as `Opts[Boolean => Unit]` instead of `Opts[Unit]`, and in main you can do:

```scala
val runOpts =
  (debugOpt,  adminCommand orElse curriculumCommands orElse propsCommands)
    .mapN { (debug, command) => command(debug) }
```

(In a real large program like yours I'll often use a trait instead of an anonymous function, to be a bit more explicit, but you get the idea hopefully.)

https://github.com/mslinn/decline-test/blob/3dce3b03e60972d36830865653f334b014f6d46d/src/main/scala/Cad.scala#L201

Note that `CommandApp` adds a help flag automatically, which you can disable with `helpFlag = false`.

Assuming you want to keep your own help flag, the end of your main might look like.

```scala
   val helpOpt = Opts.flag("help",  short = "h", help = "Displays man page.").map { _ => run("man", "cad")

  helpOpt orElse runOpts
```

**Comment by mslinn** [comment]
@bkirwi Wow, thanks for such a detailed response!

I suppose subcommands need to be defined as `Opts[Boolean => Unit]` instead of `Opts[Unit]` for short-circuiting parsing when a subcommand matches? Anyway, when I make that change the expressions I wrote now cause a predictable compile error:

```scala
val adminDbCmd2: Opts[Boolean => Unit] = Opts.subcommand("database", "Alias for 'admin db'.") {
  val database: Opts[String] = Opts.argument[String](metavar = "database")
  Opts.unit.map { _ =>
    println(s"TODO Execute 'cad admin db $database'")
  }
}
```

`Error:(24, 118) type mismatch;
 found   : Boolean => Unit
 required: com.monovore.decline.Opts[Boolean => Unit]
    val adminDbCmd2: Opts[Boolean => Unit] = Opts.subcommand("database", "Alias for 'admin db'.") { boolean: Boolean =>`

It is not obvious to me what the fix might be.

**Comment by bkirwi** [comment]
> I suppose subcommands need to be defined as Opts[Boolean => Unit] instead of Opts[Unit] for short-circuiting parsing when a subcommand matches?

Nope, subcommands can hold any type generally. (And all the short-circuiting is handled in `orElse` etc.)

The reason to have it hold a function is so that later we can pass the debug flag in. So you'd want something more or less like:

```scala
val adminDbCmd2: Opts[Boolean => Unit] = Opts.subcommand("database", "Alias for 'admin db'.") {
  Opts.argument[String](metavar = "database").map { db =>
    { debug: Boolean =>
      if (debug) println(s"Called with ${db}")
      ...
    }
  }
}
````

(You're getting that type error because you're effectively building a nested `Opts`... `Opts[Opts[...]]`. Which is always a sign of a bug.

> I added a } to the end of the line containing helpOpt definition, and return false

The `}` is necessary but the `false` is not. You want the return type unit because...

> Where should this go?

Just at the end of the `main` block. The type of `main` is `Opts[Unit]`; `helpOpt` has type `Opts[Unit]` if you drop the false, and `runOpts` has type `Opts[Unit]` because you made it by joining a `Opts[Boolean => Unit]` to an `Opts[Boolean]`, so you end up with the right final type.

**Comment by mslinn** [comment]
Again, many thanks for the help. I've checked in what I've got so far.

1. I've got the top-level help working if the man page is available:
   ```
   cad -h
   ```
   However if `man cad` fails I'd like to display `decline`'s computed help text. How might I do that? I embedded a TODO on line 26 for that.

2. Subcommand help is kinda sorta working:
   ```
   $ cad lecture -h
   Missing expected positional argument!
   
   Usage: cad lecture --help <id> <action>
   
   This Cadenza Client curriculum management command provides lecture-level authoring support.
   
   Options and flags:
       --help
           Display this help text.
       --help, -h
           Displays lecture man page.
   ```
   Two things could be better when the user requests subcommand help:
     a. Avoid the `missing positional argument` message
     b. Only show the `--help` option once in the help message.

3. The `section` command on lines 157-169 shows one way of writing the parsing incantation.
The `lecture` command on lines 171-182 shows another way of writing the parsing incantation.
Comments?

**Comment by bkirwi** [comment]
If you have a command, you can get the full help text for that command via `Help.fromCommand(command)`. I think you should be able to figure it out from there!

**Comment by mslinn** [comment]
@bkirwi My hope in engaging in this public discourse is to provide a non-trivial (and very real) example of how to use Decline. From now forward, however I expect the work on this project to become proprietary and of little incremental interest to aspiring Decline programmers. I could fork this project now, leaving this public version, and accept PRs from others, or I could transfer this project to you folks, and you could do the maintenance. As of today, I am happy either way. Please let me know your desires.

**Comment by mslinn** [comment]
I'm having trouble with `Help.fromCommand(command)`, where I have nested subcommands. How would I organize the code so subcommands could be passed? For example: https://github.com/mslinn/decline-test/blob/master/src/main/scala/AdminCmd.scala

**Comment by bkirwi** [comment]
I'm not totally sure I understand the behaviour you're looking for... is the idea just that you want to override the behaviour of `--help` at every level, so that `cad admin db --help` will also do the man page thing?

If so... the `--help` flag is added in the `Opts.subcommand` call, so if you're going to want this _everywhere_ the easiest thing to do might be to add a helper method:

```scala
def subcommand(name: String, help: String)(opts: Opts[Boolean => Unit]) = {
  lazy val help: Opts[Boolean => Unit] =
    Opts.flag("help", "...").map { _ =>
      debug => System.err.println(Help.fromCommand(command)) // or whatever
    }
  lazy val command: Command[Boolean => Unit] =
    Command(name, help, helpFlag = false) { help orElse opts }
  Opts.subcommand(command)
}
```

Which would give you a single place to add this behaviour.

> I could fork this project now, leaving this public version, and accept PRs from others, or I could transfer this project to you folks, and you could do the maintenance. As of today, I am happy either way. Please let me know your preference.

Good question! I think this is a cool example of `decline` "in the wild," and it sounds useful to keep it up as an example -- but I personally am unlikely to find the time to maintain it! Which means your first suggestion sounds good to me if it's reasonable for you.

**Comment by mslinn** [comment]
Wow, the writing style required by `decline` is unusual. I almost grasp the approach. There would be value in explaining how it works. I really appreciate the time you spent with me :)

I've improved your `subcommand` method slightly, and it took care of the outstanding issues while creating [one last issue](https://github.com/mslinn/decline-test/blob/help/src/main/scala/Cad.scala#L16).
